ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs

```diff
@@ -0,0 +1,313 @@
+{-# LANGUAGE DeriveFunctor         #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+
+-- | Error policies, and integration with 'SuspendDecision'-semigroup action on
+-- 'PeerState'.
+--
+module Ouroboros.Network.ErrorPolicy
+  ( ErrorPolicies (..)
+  , ConnectionException (..)
+  , ApplicationException (..)
+  , nullErrorPolicies
+  , ErrorPolicy (..)
+  , evalErrorPolicy
+  , evalErrorPolicies
+  , CompleteApplication
+  , Result (..)
+  , completeApplicationTx
+
+  -- * Traces
+  , ErrorPolicyTrace (..)
+  , traceErrorPolicy
+  , WithAddr (..)
+
+  -- * Re-exports of PeerState
+  , PeerStates
+  , SuspendDecision (..)
+  ) where
+
+import           Control.Exception (Exception, SomeException (..))
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Foldable (traverse_)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromMaybe, mapMaybe)
+import           Data.Semigroup (sconcat)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable ( Proxy (..)
+                               , gcast
+                               , tyConName
+                               , typeRepTyCon
+                               , typeRep
+                               )
+import           Text.Printf
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Tracer
+
+import           Data.Semigroup.Action
+
+import           Ouroboros.Network.Subscription.PeerState
+
+data ErrorPolicy e where
+     ErrorPolicy :: forall e err.
+                      Exception err
+                   => (e err -> SuspendDecision DiffTime)
+                   -> ErrorPolicy e
+
+instance Show (ErrorPolicy e) where
+    show (ErrorPolicy (_err :: e err -> SuspendDecision DiffTime)) =
+           "ErrorPolicy ("
+        ++ tyConName (typeRepTyCon (typeRep (Proxy :: Proxy err)))
+        ++ ")"
+
+
+evalErrorPolicy :: forall et e.
+                   Exception e
+                => et e
+                -> ErrorPolicy et
```

Comment:
Hmm, I would try to be consistent with naming the type parameters (`et`, `e`, `err`). In this function, you use `e` for the exception type and `et` for `ErrorPolicy`'s `e` parameter, whereas in `ErrorPolicy` you have `err` for the exception type and `e` as a wrapper.
I don't kind care much about `et` vs. `e` vs. `err`, I just want it to be consistent :slightly_smiling_face:.

This comment applies to the whole file (and more), not just to this function in particular.

Reply:
I will change to use `et` for exception type and e for exception everywhere.